src/NUnitConsole/nunit-console/Utilities/TeamCityAssemblyResolver.cs

```diff
@@ -0,0 +1,121 @@
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace NUnit.ConsoleRunner.Utilities
+{
+    internal class TeamCityAssemblyResolver
+    {
+        private static readonly char[] Separator = { '-' };
+        private readonly object lockObject = new object();
+        private readonly Dictionary<string, string> _assemblyNames = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
+
+        public void RegisterSuite(string suiteId, string fullName)
+        {
+            string flowId;
+            string assemblyName;
+            if (!TryGetFlowId(suiteId, out flowId) || !TryGetAssemblyName(fullName, out assemblyName))
+            {
+                return;
+            }
+
+            lock (lockObject)
+            {
+                this._assemblyNames[flowId] = assemblyName;
+            }
+        }
+
+        public void UnregisterSuite(string suiteId, string fullName)
+        {
+            string flowId;
+            string assemblyName;
+            if (!TryGetFlowId(suiteId, out flowId) || !TryGetAssemblyName(fullName, out assemblyName))
+            {
+                return;
+            }
+
+            lock (lockObject)
+            {
+                this._assemblyNames.Remove(flowId);
+            }
+        }
+
+        public bool TryResolveAssembly(string testId, out string assemblyName)
+        {
+            string flowId;
+            if (!TryGetFlowId(testId, out flowId))
+            {
+                assemblyName = default(string);
+                return false;
+            }
+
+            lock (lockObject)
+            {
+                return this._assemblyNames.TryGetValue(flowId, out assemblyName);
+            }
+        }
+
+        internal static bool TryGetFlowId(string id, out string flowId)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                flowId = default(string);
+                return false;
+            }
+
+            var idItems = id.Split(Separator, StringSplitOptions.RemoveEmptyEntries);
+            if (idItems.Length != 2 || string.IsNullOrEmpty(idItems[0].Trim()))
+            {
+                flowId = default(string);
+                return false;
+            }
+
+            flowId = idItems[0];
```

Comment:
I suggested that the first part of test id should be unique for the assembly and for each test within this assembly. Thus this the id could be used to get the assembly name